src/ducks/select/index.js

```diff
@@ -69,18 +71,34 @@ const reducer = (state = INITIAL_STATE, action) => {
 };
 
 function applyToggleItem(state, action) {
-  let { stateKey, id } = action.payload;
-  let list = state[stateKey] || [];
-  let index = list.indexOf(id);
-  let selectedItems;
-
-  if (index !== -1) {
-    selectedItems = removeItem(list, index);
-  } else {
-    selectedItems = addItem(list, id);
+  const { stateKey, id, event, allIds } = action.payload;
+  const currentSelection = state[stateKey] && state[stateKey].selectedItems ?
+    state[stateKey].selectedItems :
+    [];
+  const index = currentSelection.indexOf(id);
+  const isAlreadySelected = index !== -1;
+  if (!isAlreadySelected && allIds && allIds.length && event && event.shiftKey) {
```

Comment:
Rather than using the same action (doSelectItem), I'd introduce a new and let the component do the `if (shiftKey) doRangeSelect else doNormalSelect`. Here we are coupling the model to the user interface, so we're mixing concerns. That the shift key (or an Event in general) is involved is just an user interface detail.

An alternative could be to use doSelectItems and let the component figure out which items the user meant. Perhaps nicer, since we're then keeping the model/redux simpler. But I'd go for whatever is easier.